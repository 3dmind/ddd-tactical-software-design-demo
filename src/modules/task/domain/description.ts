import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface DescriptionProps {
  value: string;
}

export class Description extends ValueObject<DescriptionProps> {
  public static minLength = 2;
  public static maxLength = 250;

  private constructor(props: DescriptionProps) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(text: string): Result<Description> {
    const nullGuardResult = Guard.againstNullOrUndefined(text, 'Description');

    if (!nullGuardResult.succeeded) {
      return Result.fail<Description>(nullGuardResult.message);
    }

    const minGuardResult = Guard.againstAtLeast(this.minLength, text);
    const maxGuardResult = Guard.againstAtMost(this.maxLength, text);

    if (!minGuardResult.succeeded) {
      return Result.fail<Description>(minGuardResult.message);
    }

    if (!maxGuardResult.succeeded) {
      return Result.fail<Description>(maxGuardResult.message);
    }

    return Result.ok<Description>(new Description({ value: text }));
  }
}
