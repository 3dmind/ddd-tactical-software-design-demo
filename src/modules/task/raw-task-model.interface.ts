export interface RawTaskModelInterface {
  archived: boolean;
  archivedAt: Date;
  createdAt: Date;
  description: string;
  discarded: boolean;
  discardedAt: Date;
  editedAt: Date;
  resumedAt: Date;
  taskId: string;
  tickedOff: boolean;
  tickedOffAt: Date;
}
