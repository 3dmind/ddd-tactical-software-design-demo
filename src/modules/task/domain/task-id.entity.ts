import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class TaskIdEntity extends Entity<any> {
  private constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<TaskIdEntity> {
    return Result.ok<TaskIdEntity>(new TaskIdEntity(id));
  }
}
