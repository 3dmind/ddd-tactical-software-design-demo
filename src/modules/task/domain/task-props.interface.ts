import { DescriptionValueObject } from './description.value-object';

export interface TaskPropsInterface {
  archived: boolean;
  archivedAt: Date;
  createdAt: Date;
  description: DescriptionValueObject;
  resumedAt: Date;
  tickedOff: boolean;
  tickedOffAt: Date;
}
