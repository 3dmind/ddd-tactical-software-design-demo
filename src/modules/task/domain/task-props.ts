import { Description } from './description';

export interface TaskProps {
  description: Description;
  createdAt: Date;
  tickedOff: boolean;
  tickedOffAt?: Date;
}
