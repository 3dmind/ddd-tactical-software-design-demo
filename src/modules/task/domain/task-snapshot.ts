import { Description } from './description';
import { TaskId } from './task-id';
import { TaskProps } from './task-props';

interface TaskSnapshotProps {
  readonly createdAt: Date;
  readonly description: Description;
  readonly isTickedOff: boolean;
  readonly taskId: TaskId;
}

export class TaskSnapshot implements TaskSnapshotProps {
  readonly createdAt: Date;
  readonly description: Description;
  readonly isTickedOff: boolean;
  readonly taskId: TaskId;

  constructor(props: Readonly<TaskProps>, taskId: TaskId) {
    this.createdAt = props.createdAt;
    this.description = props.description;
    this.isTickedOff = props.tickedOff;
    this.taskId = taskId;

    Object.freeze(this);
  }
}
