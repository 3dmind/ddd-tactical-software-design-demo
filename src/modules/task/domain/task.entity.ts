import { Guard, Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';
import { DescriptionValueObject } from './description.value-object';
import { TaskIdEntity } from './task-id.entity';
import { TaskPropsInterface } from './task-props.interface';
import { TaskSnapshot } from './task-snapshot';

export class TaskEntity extends Entity<TaskPropsInterface> {
  private constructor(props: TaskPropsInterface, id?: UniqueEntityId) {
    super(props, id);
  }

  get taskId(): TaskIdEntity {
    return TaskIdEntity.create(this._id).getValue();
  }

  public static create(
    props: TaskPropsInterface,
    id?: UniqueEntityId,
  ): Result<TaskEntity> {
    const nullGuard = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.description,
        argumentName: 'description',
      },
      {
        argument: props.createdAt,
        argumentName: 'createdAt',
      },
      {
        argument: props.tickedOff,
        argumentName: 'tickedOff',
      },
    ]);

    if (!nullGuard.succeeded) {
      return Result.fail<TaskEntity>(nullGuard.message);
    }

    return Result.ok<TaskEntity>(new TaskEntity(props, id));
  }

  public static note(description: DescriptionValueObject): Result<TaskEntity> {
    return TaskEntity.create({
      createdAt: new Date(),
      description,
      resumedAt: null,
      tickedOff: false,
      tickedOffAt: null,
    });
  }

  createSnapshot(): TaskSnapshot {
    return new TaskSnapshot(this.props, this.taskId);
  }

  tickOff(): void {
    this.props.tickedOff = true;
    this.props.tickedOffAt = new Date();
  }

  isTickedOff(): boolean {
    return this.props.tickedOff;
  }

  resume(): void {
    this.props.tickedOff = false;
    this.props.resumedAt = new Date();
  }
}
