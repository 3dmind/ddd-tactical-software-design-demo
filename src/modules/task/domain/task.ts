import { Guard, Result } from '../../../shared/core';
import { Entity, UniqueEntityID } from '../../../shared/domain';
import { Description } from './description';
import { TaskId } from './task-id';
import { TaskProps } from './task-props';
import { TaskSnapshot } from './task-snapshot';

export class Task extends Entity<TaskProps> {
  private constructor(props: TaskProps, id?: UniqueEntityID) {
    super(props, id);
  }

  get taskId(): TaskId {
    return TaskId.create(this._id).getValue();
  }

  public static create(props: TaskProps, id?: UniqueEntityID): Result<Task> {
    const nullGuard = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.description,
        argumentName: 'description',
      },
      {
        argument: props.createdAt,
        argumentName: 'createdAt',
      },
      {
        argument: props.tickedOff,
        argumentName: 'tickedOff',
      },
    ]);

    if (!nullGuard.succeeded) {
      return Result.fail<Task>(nullGuard.message);
    }

    return Result.ok<Task>(new Task(props, id));
  }

  public static note(description: Description): Result<Task> {
    return Task.create({
      description,
      createdAt: new Date(),
      tickedOff: false,
    });
  }

  createSnapshot(): TaskSnapshot {
    return new TaskSnapshot(this.props, this.taskId);
  }
}
