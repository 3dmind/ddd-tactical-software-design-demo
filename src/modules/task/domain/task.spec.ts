import * as faker from 'faker';
import * as uuid from 'uuid';
import { UniqueEntityID } from '../../../shared/domain';
import { Description } from './description';
import { Task } from './task';
import { TaskId } from './task-id';

jest.mock('uuid');

describe('Task', () => {
  beforeAll(() => {
    uuid.v4.mockReturnValue(faker.random.uuid());
  });

  afterAll(() => {
    uuid.mockRestore();
  });

  it(`should guard 'description' against 'null' or 'undefined'`, () => {
    const taskResultNull = Task.create({
      createdAt: null,
      description: null,
      tickedOff: null,
    });
    const taskResultUndefined = Task.create({
      createdAt: undefined,
      description: undefined,
      tickedOff: undefined,
    });

    expect(taskResultNull.isFailure).toBe(true);
    expect(taskResultUndefined.isFailure).toBe(true);
  });

  it('should create task', () => {
    const text = faker.lorem.words(5);
    const entityId = new UniqueEntityID();
    const description = Description.create(text).getValue();

    const taskResult = Task.create(
      {
        createdAt: new Date(),
        description,
        tickedOff: false,
      },
      entityId,
    );
    const task = taskResult.getValue();

    expect(taskResult.isSuccess).toBe(true);
    expect(task.taskId).toBeInstanceOf(TaskId);
    expect(task.taskId.id.equals(entityId)).toBe(true);
  });

  it('should note new task', () => {
    const text = faker.lorem.words(5);
    const description = Description.create(text).getValue();

    const taskResult = Task.note(description);
    const task = taskResult.getValue();

    expect(taskResult.isSuccess).toBe(true);
    expect(task.taskId.id.toValue()).toBeDefined();
  });

  it('should take a snapshot of its internal state', () => {
    const uuid = '8597ccd9-4237-44a6-b434-8836693c4b51';
    const text = 'Lorem ipsum';
    const entityId = new UniqueEntityID(uuid);
    const description = Description.create(text).getValue();
    const task = Task.create(
      {
        createdAt: new Date(Date.parse('1977-01-01')),
        description,
        tickedOff: false,
      },
      entityId,
    ).getValue();

    const taskSnapshot = task.createSnapshot();

    expect(taskSnapshot).toMatchSnapshot();
    expect(Object.isFrozen(taskSnapshot)).toBe(true);
  });
});
