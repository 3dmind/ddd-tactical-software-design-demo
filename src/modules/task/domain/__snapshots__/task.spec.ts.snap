// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task should take a snapshot of its internal state 1`] = `
TaskSnapshot {
  "createdAt": 1977-01-01T00:00:00.000Z,
  "description": Description {
    "props": Object {
      "value": "Lorem ipsum",
    },
  },
  "isTickedOff": false,
  "taskId": TaskId {
    "_id": UniqueEntityID {
      "value": "8597ccd9-4237-44a6-b434-8836693c4b51",
    },
    "props": null,
  },
}
`;
